ansible_base/api_documentation/postprocessing_hooks.py

```diff
@@ -0,0 +1,466 @@
+    for prep in prepositions:
+        if prep in purpose:
+            # Split on first occurrence to separate noun phrase from context
+            parts = purpose.split(prep, 1)
+            noun_phrase = parts[0]
+            context = parts[1] if len(parts) > 1 else ''
+
+            # Singularize the noun phrase (last word)
+            words = noun_phrase.split()
+            if words:
+                last_word = words[-1]
+                singular_last = singularize_resource(last_word)
+                words[-1] = singular_last
+                singular_noun = ' '.join(words)
+            else:
+                singular_noun = noun_phrase
+
+            # Reconstruct with preposition
+            return f"{singular_noun}{prep}{context}"
+
+    # No preposition found, singularize the last word of the entire string
+    words = purpose.split()
+    if not words:
+        return purpose
+
+    last_word = words[-1]
+    singular_last = singularize_resource(last_word)
+    words[-1] = singular_last
+    return ' '.join(words)
+
+
+def generate_associate_description(operation_id, path, resource_name):
+    """Generate description for associate/disassociate operations."""
+    is_associate = operation_id.endswith('_associate_create')
+
+    # Filter out API prefixes (everything up to version string) using generic approach
+    path_parts = filter_api_prefixes(parse_path_segments(path))
+
+    # Remove associate/disassociate if it's the last segment (special action marker)
+    if path_parts and path_parts[-1] in ['associate', 'disassociate']:
+        path_parts = path_parts[:-1]
+
+    if len(path_parts) >= 2:
+        child = path_parts[-1].replace('_', ' ')
+        parent_singular = singularize_resource(path_parts[-2])
+        parent_readable = parent_singular.replace('_', ' ')
+        article = 'an' if parent_readable[0].lower() in 'aeiou' else 'a'
+
+        verb = "Associate" if is_associate else "Disassociate"
+        preposition = "with" if is_associate else "from"
+        return f"{verb} {child} {preposition} {article} {parent_readable}"
+
+    # Fallback if pattern doesn't match
+    verb = "Associate" if is_associate else "Disassociate"
+    return f"{verb} {resource_name}"
+
+
+def generate_crud_description(action, operation_name, resource_name, parent_resource):
+    """Generate description for standard CRUD operations."""
+    # For list operations, use the name as-is (already formatted)
+    if action == 'list':
+        return f"{operation_name} {resource_name}"
+
+    # For partial_update, use special formatting
+    if action == 'partial_update':
+        if parent_resource:
+            return f"Partially update existing {resource_name}"
+        return f"Partially update existing {singularize_resource(resource_name)}"
+
+    # For other CRUD operations (retrieve, read, create, update, destroy, delete)
+    if action in ['retrieve', 'read', 'create', 'update', 'destroy', 'delete']:
```

Comment:
Could this be something more like `OPERATION_DESCRIPTIVE_NAMES.keys()` or somehow otherwise be auto-generated instead of a separate maintained list (applies to other locations).